src/collapse/common.rs

```diff
@@ -0,0 +1,780 @@
+                "$LP$" => "(",
+                "$RP$" => ")",
+                "$C$" => ",",
+                "$u7e$" => "~",
+                "$u20$" => " ",
+                "$u27$" => "'",
+                "$u3d$" => "=",
+                "$u5b$" => "[",
+                "$u5d$" => "]",
+                "$u7b$" => "{",
+                "$u7d$" => "}",
+                "$u3b$" => ";",
+                "$u2b$" => "+",
+                "$u21$" => "!",
+                "$u22$" => "\"",
+            }
+        } else {
+            let idx = match rest.char_indices().find(|&(_, c)| c == '$' || c == '.') {
+                None => rest.len(),
+                Some((i, _)) => i,
+            };
+            demangled.push_str(&rest[..idx]);
+            rest = &rest[idx..];
+        }
+    }
+
+    Cow::Owned(demangled)
+}
+
+#[cfg(test)]
+pub(crate) mod testing {
+    use std::collections::HashMap;
+    use std::fmt;
+    use std::fs::File;
+    use std::io::Write;
+    use std::io::{self, BufRead, Read};
+    use std::path::{Path, PathBuf};
+    use std::time::{Duration, Instant};
+
+    use libflate::gzip::Decoder;
+
+    use super::*;
+    use crate::collapse::Collapse;
+
+    pub(crate) trait DurationExt {
+        fn as_nanos_compat(&self) -> u128;
+    }
+
+    impl DurationExt for Duration {
+        fn as_nanos_compat(&self) -> u128 {
+            self.as_secs() as u128 * 1_000_000_000 + self.subsec_nanos() as u128
+        }
+    }
+
+    pub(crate) fn read_inputs<P>(inputs: &[P]) -> io::Result<HashMap<PathBuf, Vec<u8>>>
+    where
+        P: AsRef<Path>,
+    {
+        let mut map = HashMap::default();
+        for path in inputs.iter() {
+            let path = path.as_ref();
+            let bytes = {
+                let mut buf = Vec::new();
+                let mut file = File::open(path)?;
+                if path.to_str().unwrap().ends_with(".gz") {
+                    let mut reader = Decoder::new(file)?;
+                    reader.read_to_end(&mut buf)?;
+                } else {
+                    file.read_to_end(&mut buf)?;
+                }
+                buf
+            };
+            map.insert(path.to_path_buf(), bytes);
+        }
+        Ok(map)
+    }
+
+    pub(crate) fn test_collapse_multi<C, P>(folder: &mut C, inputs: &[P]) -> io::Result<()>
+    where
+        C: Collapse,
+        P: AsRef<Path>,
+    {
+        const MAX_THREADS: usize = 16;
+        for (_, bytes) in read_inputs(inputs)? {
+            folder.set_nthreads(1);
+            let mut writer = Vec::new();
+            folder.collapse(&bytes[..], &mut writer)?;
+            let expected = std::str::from_utf8(&writer[..]).unwrap();
+
+            for n in 2..=MAX_THREADS {
+                folder.set_nthreads(n);
+                let mut writer = Vec::new();
+                folder.collapse(&bytes[..], &mut writer)?;
+                let actual = std::str::from_utf8(&writer[..]).unwrap();
+
+                assert_eq!(actual, expected);
```

Comment:
I think we should capture the filename in the `for` and then include the filename in the error message here.